src/ast/nodes/ObjectExpression.ts

```diff
@@ -38,25 +43,29 @@ export default class ObjectExpression extends NodeBase {
 	) {
 		if (path.length === 0) return;
 
-		const { properties, hasCertainHit } = this.getPossiblePropertiesWithName(
+		const { properties } = this.getPossiblePropertiesWithName(
 			path[0],
 			PROPERTY_KINDS_READ,
 			options
 		);
-		if (hasCertainHit) {
-			for (const property of properties) {
-				property.forEachReturnExpressionWhenCalledAtPath(
-					path.slice(1),
-					callOptions,
-					callback,
-					options
-				);
-			}
+		for (const property of properties) {
+			property.forEachReturnExpressionWhenCalledAtPath(
+				path.slice(1),
+				callOptions,
+				callback,
+				options
+			);
```

Comment:
In theory `hasCertainHit` should have caught this case:

```js
var x = {
	a () {
	},
	[unknown] () {
  }
};
x.a();
```

but it seems it wasn't applying to such a thing anyway so that removing this is fine.

That said, if we extend this type of logic to deal with progressive property sets within reduced control flows like:

```js
var exports = {};

exports.a = function () {
}

exports.a();
```

then we may want to be detecting this again.

Reply:
> in theory hasCertainHit should have caught this case

Not 100% sure what you mean by "caught" as I do not see any side-effects to be detected. In general one thing this algorithm tries to capture is that when two properties with the same name appear in a literal, the second overwrites the first. Thus when we access `.a` and an unknown property appears after `a`, both need to be considered while any unknown properties before `a` can be safely ignored. When there are only unknown properties, calling a property is always a side-effect: [REPL](https://rollupjs.org/repl?version=0.59.2&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMiUyRiUyRiUyMHJlbW92ZWQlNUNudmFyJTIweCUyMCUzRCUyMCU3QiU1Q24lNUN0JTVCdW5rbm93biU1RCUyMCgpJTIwJTdCJTVDbiU1Q3QlNUN0Y29uc29sZS5sb2coMSklNUNuJTIwJTIwJTdEJTJDJTVDbiU1Q3RhJTIwKCklMjAlN0IlN0QlNUNuJTdEJTNCJTVDbnguYSgpJTNCJTVDbiU1Q24lMkYlMkYlMjByZXRhaW5lZCUyMGJlY2F1c2UlMjBvbmUlMjBwb3NzaWJsZSUyMGhpdCUyMHdpdGglMjBzaWRlLWVmZmVjdHMlNUNudmFyJTIweSUyMCUzRCUyMCU3QiU1Q24lNUN0YSUyMCgpJTIwJTdCJTdEJTJDJTVDbiU1Q3QlNUJ1bmtub3duJTVEJTIwKCklMjAlN0IlNUNuJTVDdCU1Q3Rjb25zb2xlLmxvZygxKSU1Q24lMjAlMjAlN0QlMkMlNUNuJTdEJTNCJTVDbnkuYSgpJTNCJTVDbiU1Q24lMkYlMkYlMjByZXRhaW5lZCUyMGJlY2F1c2UlMjBubyUyMGNlcnRhaW4lMjBoaXQlNUNudmFyJTIweiUyMCUzRCUyMCU3QiU1Q24lNUN0JTVCdW5rbm93biU1RCUyMCgpJTIwJTdCJTdEJTVDbiU3RCUzQiU1Q256LmEoKSUzQiUyMiU3RCU1RCUyQyUyMm9wdGlvbnMlMjIlM0ElN0IlMjJmb3JtYXQlMjIlM0ElMjJlcyUyMiUyQyUyMm5hbWUlMjIlM0ElMjJteUJ1bmRsZSUyMiUyQyUyMmdsb2JhbHMlMjIlM0ElN0IlN0QlMkMlMjJhbWQlMjIlM0ElN0IlMjJpZCUyMiUzQSUyMiUyMiU3RCU3RCUyQyUyMmV4YW1wbGUlMjIlM0FudWxsJTdE)

> That said, if we extend this type of logic to deal with progressive property sets 

Yes, these cases should certainly still be considered. I think there should already be some tests for this. I have thought a little about how an SSA form could be extended to handle property reassignments. One concept that comes to my mind is considering object expressions to be like "scopes" where the properties are the "variables" that are duplicated:

```javascript
// code
const x = {a: false};
const y = x;
x.a = true;
if (y.a) console.log('properly tracked');

// "SSA form"
// the object expression OE on the right side keeps track of its properties
const x1 = {a1: false};
const y1 = x1;

// we define version 2 of property a; the actual representation will probably be different
// note that this does NOT define a new x2 as x still refers to the same object reference OE
x1.a2 = true;

// as we know that y1 refers to OE, we know we need to use version 2 here
if (y1.a2) console.log('properly tracked');

// tree-shaken result
console.log('properly tracked');
```

You can already see that something like this needs very sophisticated analysis and for the first attempt at SSA I would not include property assignments but focus on actual variable reassignments.

Reply:
In theory, `unknown` could be equal to `'a'`, which would then take precedence meaning it is not a certain hit even though there is an explicit `'a'` defined. certainHit logic has precedence logic in it (loop in reverse order), yet doesn't quite go all the way in taking these concepts to their complete conclusions.

We lose the ability to go back on this here, but yes it's a super edge case. I'm mentioning because we are so close to handling the edge case, and now moving away from it. Projects that get massive traction need to be watertight - that is a big difference between a useful software project, and a massively supported and used project.

Reply:
Very interesting object interpretation of SSA.

Somewhat unrelated, but I've long wanted to have the available analysis to form the following optimization:

```js
var x = 'asdf';
var z = 'g';

export function y () {
  console.log(z);
}

export function p() {
  console.log(x);
}
```

If one module uses just `y` and another module uses just `p`, we don't need to keep the definitions of y and p in the same chunk necessarily and can actually split the above module into two separate modules:

```js
var x = 'asdf';
export function p() {
  console.log(x);
}
```

```js
var z = 'g';
export function y () {
  console.log(z);
}
```

This is a super complex long-term code splitting optimization I've wanted to implement for a while, and likely won't be able to implement for a while - but mentioning here as it would usefully piggy-back off similar analysis.

Reply:
> In theory, unknown could be equal to 'a', which would then take precedence meaning it is not a certain hit even though there is an explicit 'a' defined. certainHit logic has precedence logic in it (loop in reverse order), yet doesn't quite go all the way in taking these concepts to their complete conclusions

Actually as far as I know, it does go all the way 😜 Let me elaborate:
`hasCertainHit` means "we are certain there is a property of the given name". It does *not* mean we are certain which one it is! And in fact it is only used in the negative: When we are *not* certain there is a property of the given name, calling it must always be a side-effect.

If we *are* certain on the other hand, we check a whole array of properties for side-effects. Cf. this example: [REPL](https://rollupjs.org/repl?version=0.59.2&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmNvbnN0JTIweCUyMCUzRCUyMCU3QiU1Q24lNUN0YSgpJTIwJTdCJTVDbiU1Q3QlMjAlMjB0cmlnZ2VyZWQlMjAlM0QlMjB0cnVlJTIwJTJGJTJGJTIwdGhpcyUyMHNpZGUtZWZmZWN0JTIwaXMlMjBjb21wbGV0ZWx5JTIwaWdub3JlZCU1Q24lNUN0JTdEJTJDJTVDbiU1Q3RhKCklMjAlN0IlN0QlMkMlNUNuJTVDdCU1QnVua25vd24lNUQoKSUyMCU3QiU1Q24lNUN0JTVDdHRyaWdnZXJlZCUyMCUzRCUyMHRydWUlM0IlMjAlMkYlMkYlMjB0cnklMjByZW1vdmluZyUyMHRoaXMlMjBzaWRlLWVmZmVjdCElNUNuJTVDdCU3RCU1Q24lN0QlM0IlNUNuJTVDbmxldCUyMHRyaWdnZXJlZCUyMCUzRCUyMGZhbHNlJTNCJTVDbnguYSgpJTNCJTVDbmFzc2VydC5vayh0cmlnZ2VyZWQpJTNCJTVDbiUyMiU3RCU1RCUyQyUyMm9wdGlvbnMlMjIlM0ElN0IlMjJmb3JtYXQlMjIlM0ElMjJlcyUyMiUyQyUyMm5hbWUlMjIlM0ElMjJteUJ1bmRsZSUyMiUyQyUyMmdsb2JhbHMlMjIlM0ElN0IlN0QlMkMlMjJhbWQlMjIlM0ElN0IlMjJpZCUyMiUzQSUyMiUyMiU3RCU3RCUyQyUyMmV4YW1wbGUlMjIlM0FudWxsJTdE)

Here we do have a certain hit; the algorithm, however, does not only check the last method 
 with name `"a"` for side-effects but also all computed properties after it.

I have added this example + some variations as a function test to make sure this works even after the update and keeps working in the future.

> Somewhat unrelated, but I've long wanted to have the available analysis to form the following optimization

Doing this in an arbitrary way might be difficult if the right hand side of the declaration itself depends on other variables. Something simpler that might be able to solve this problem in some situations and that I would consider quite doable would be inlining variables that are only used in one place (if they are used in more places, you would loose the reference identity). If done right i.e. in a way that "a variable can be inlined if it is only used once or only depends on other variables that can be inlined", then  this would inline .e.g `z` into `y` and move both into the chunk where they are used

Reply:
What I mean by being inconsistent is that the example is tree-shaken - https://rollupjs.org/repl?version=0.59.2&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMnZhciUyMHglMjAlM0QlMjAlN0IlNUNuJTVDdGElMjAoKSUyMCU3QiU1Q24lNUN0JTdEJTJDJTVDbiU1Q3QlNUJ1bmtub3duJTVEJTIwKCklMjAlN0IlNUNuJTIwJTIwJTdEJTVDbiU3RCUzQiU1Q254LmEoKSUzQiUyMiU3RCU1RCUyQyUyMm9wdGlvbnMlMjIlM0ElN0IlMjJmb3JtYXQlMjIlM0ElMjJjanMlMjIlMkMlMjJuYW1lJTIyJTNBJTIybXlCdW5kbGUlMjIlMkMlMjJnbG9iYWxzJTIyJTNBJTdCJTdEJTJDJTIyYW1kJTIyJTNBJTdCJTIyaWQlMjIlM0ElMjIlMjIlN0QlN0QlMkMlMjJleGFtcGxlJTIyJTNBbnVsbCU3RA== when in theory it shouldn't necessarily be.

> Doing this in an arbitrary way might be difficult if the right hand side of the declaration itself depends on other variables. Something simpler that might be able to solve this problem in some situations and that I would consider quite doable would be inlining variables that are only used in one place (if they are used in more places, you would loose the reference identity). If done right i.e. in a way that "a variable can be inlined if it is only used once or only depends on other variables that can be inlined", then this would inline .e.g z into y and move both into the chunk where they are used

I feel quite strongly on this one that "gathering related branches" is an important part of the module splitting process.

Reply:
> when in theory it shouldn't necessarily be

In which theory? Add a side-effect and it will not be removed. Do you want the algorithm to be dumber in this place? Why?